lib/sms/sms_submit.c

```diff
@@ -88,6 +104,102 @@ static int sms_submit_encode_number(
 	return 0;
 }
 
+/**
+ * @brief Encode SMS Service Center Address (SCA) field.
+ *
+ * @details SCA field is used in SMS-SUBMIT message. This is specified piece by piece
+ * in the following specifications:
+ *   - 3GPP TS 23.040 chapter 9.1.2.5
+ *   - 3GPP TS 27.005 chapter 3.1 (see <pdu>, <tosca> and <toda>)
+ *   - 3GPP TS 27.005 chapter 3.3.1
+ *   - 3GPP TS 27.005 chapter 4.3
+ *
+ * @param[out] sca_field Buffer for SCA field.
+ *                       Must be at least 23 bytes long to include the following information:
+ *                       - Address/number: 20
+ *                       - Address length: 1
+ *                       - Type of address: 1
+ *                       - NUL termination: 1
+ */
+static void sms_submit_encode_sca(char *sca_field)
```

Comment:
This function (which is most of this PR) takes flash as follows for configurations not using CSCA, i.e., not include gcf_sms library and using SMS as earlier:
* 760 bytes with `CONFIG_SMS_LOG_LEVEL_DBG=y`,
* 432 bytes with default logging, that is, `CONFIG_SMS_LOG_LEVEL_INF=y`
* 320 bytes with `CONFIG_LOG=n`

I was thinking of making a Kconfig for `AT+CSCA` usage to optimize flash usage but I'm hesitant given it's only 320 bytes.